### PR TITLE
Add Muon optimizer

### PR DESCRIPTION
**Pull request by Adhithya-Laxman**
## Description

This PR implements the **Muon optimizer** using pure NumPy, completing the neural network optimizers module for the repository.

Muon is a cutting-edge optimizer specifically designed for hidden layer weight matrices in neural networks, using Newton-Schulz matrix orthogonalization iterations for improved convergence and computational efficiency.

**This PR addresses part of issue #13662** - [Add neural network optimizers module to enhance training capabilities](https://github.com/TheAlgorithms/Python/issues/13662)

## What does this PR do?

- Implements Muon optimizer that applies Newton-Schulz orthogonalization to gradient matrices
- Specifically optimized for 2D weight matrices (hidden layers)
- Provides superior convergence properties compared to standard optimizers
- Uses momentum with orthogonalized gradient updates
- Memory-efficient implementation with minimal state tracking
- Educational implementation suitable for learning advanced optimization techniques

## Implementation Details

- **Algorithm**: Muon (Matrix-aware Optimizer with Orthogonalization)
- **Core Mechanism**: Newton-Schulz iterations for gradient orthogonalization
- **Update rule**: 
  ```
  ortho_grad = NewtonSchulz(gradient, steps=5)
  velocity = momentum * velocity + ortho_grad
  param = param - learning_rate * velocity
  ```
- **Newton-Schulz Iteration**:
  ```
  For k steps:
    A = 1.5 * X - 0.5 * X @ (X.T @ X)
  Returns orthogonalized matrix
  ```
- **Key Features**:
  - Matrix-aware optimization
  - Spectral norm-based updates
  - Implicit regularization through orthogonalization
  - Momentum accumulation with orthogonalized gradients
- **Pure NumPy**: No PyTorch, TensorFlow, or other frameworks required
- **Educational focus**: Clear implementation showing matrix orthogonalization concepts

## Why Muon?

Muon represents state-of-the-art optimizer research and offers several advantages:

- **Computational Efficiency**: ~2x more efficient than AdamW in large-scale training
- **Better Convergence**: Particularly effective for training deep networks
- **Memory Efficient**: Maintains only momentum buffer (no second-order moments)
- **Geometric Awareness**: Exploits matrix structure of neural network parameters
- **Recent Innovation**: Used in current training speed records for NanoGPT and CIFAR-10

## Features

✅ Complete docstrings with parameter descriptions  
✅ Type hints for all function parameters and return values  
✅ Doctests for correctness validation  
✅ Usage example demonstrating optimizer on matrix optimization  
✅ PEP8 compliant code formatting  
✅ Newton-Schulz orthogonalization implementation  
✅ Configurable hyperparameters (learning rate, momentum, iteration steps)  
✅ Pure NumPy - no external deep learning frameworks  

## Testing

All doctests pass:
```
python -m doctest neural_network/optimizers/muon.py -v
```

Linting passes:
```
ruff check neural_network/optimizers/muon.py
```

Example output demonstrates proper optimization behavior on matrix parameters.

## References

- [Muon: An optimizer for hidden layers (Original Blog Post)](https://kellerjordan.github.io/posts/muon/)
- [Muon GitHub Repository](https://github.com/KellerJordan/Muon)
- [Muon is Scalable for LLM Training (arXiv)](https://arxiv.org/abs/2502.16982)
- [Practical Efficiency of Muon for Pretraining (arXiv)](https://arxiv.org/abs/2505.02222)
- [Newton-Schulz Iteration - Matrix Functions](https://en.wikipedia.org/wiki/Matrix_function)

## Relation to Issue #13662

This PR completes the optimizer sequence outlined in #13662:
- ✅ Stochastic Gradient Descent (SGD) - #13671
- ✅ Momentum SGD - #13680
- ✅ Nesterov Accelerated Gradient (NAG) - #13718
- ✅ Adagrad - #13681
- ✅ Adam - #13718
- ✅ **Muon** (this PR) - **FINAL OPTIMIZER**

With this PR, the neural network optimizers module is now complete with 6 fundamental optimizers covering classical to cutting-edge optimization techniques.

## Use Cases

Muon is particularly effective for:
- Training large language models and transformers
- Deep convolutional networks
- Scenarios requiring computational efficiency at scale
- Research into matrix-aware optimization
- Cases where AdamW shows diminishing returns

## Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md)
- [x] This pull request is all my own work -- I have not plagiarized
- [x] I know that pull requests will not be merged if they fail the automated tests
- [x] This PR only changes one algorithm file
- [x] All new Python files are placed inside an existing directory
- [x] All filenames are in all lowercase characters with no spaces or dashes
- [x] All functions and variable names follow Python naming conventions
- [x] All function parameters and return values are annotated with Python type hints
- [x] All functions have doctests that pass the automated testing
- [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation

## Summary

This PR marks the completion of the neural network optimizers module, providing educators and learners with a comprehensive collection of optimization algorithms from fundamental SGD to cutting-edge Muon. The module now serves as a complete educational resource for understanding neural network training optimization.

---

## This PR along with the following PRs collectively fixes issue #13662:

**Related PRs:**
- [feat: add SGD optimizer for neural networks #13671](https://github.com/TheAlgorithms/Python/pull/13671)
- [Add Momentum SGD optimizer implementation #13680](https://github.com/TheAlgorithms/Python/pull/13680)
- [Add Adagrad optimizer implementation in Pure Numpy #13681](https://github.com/TheAlgorithms/Python/pull/13681)
- [Added Nesterov and Adam Optimizers #13718](https://github.com/TheAlgorithms/Python/pull/13718)
- **This PR (Muon optimizer)** - Final optimizer completing the module

**Fixes #13662** (This PR completes the neural network optimizers module)

### CODE REVIEW

**Comment by Adhithya-Laxman** [comment]
My other PRs are closed now, Kindly help me in opening a PR for this branch. 
This is the final branch that fixes the issue
